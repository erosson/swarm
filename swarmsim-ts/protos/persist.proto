syntax = "proto3";
package swarmsim.persist;
import "google/protobuf/timestamp.proto";

message Persist {
  // TODO there's probably a better way to encode decimals, and these maps...
  /**
   * Map of {unittype.protoId: count-as-Decimal.js}
   */
  map<uint32, Decimal> units = 1;
  /**
   * Map of {upgrade.protoId: count-as-Decimal.js}
   */
  map<uint32, Decimal> upgrades = 2;
  /**
   * Map of {achievement.protoId: elapsed-ms}
   */
  map<uint32, double> achievements = 3;
  /**
   * Map of {upgrade.protoId: [-1,0,1,2,4]}
   */
  map<uint32, int32> watched = 4;
  Statistics statistics = 5;
  Options options = 6;
  optional AchievementsUI achievementsUI = 7;

  Version version = 20;
  google.protobuf.Timestamp started = 21;
  google.protobuf.Timestamp closed = 22;
  google.protobuf.Timestamp reified = 23;
  google.protobuf.Timestamp restarted = 24;
  optional google.protobuf.Timestamp saved = 25;
  optional google.protobuf.Timestamp crystalCooldownDoneAt = 26;
  double skippedMillis = 27;
}

message Version {
  string saved = 1;
  string started = 2;
}

message Statistics {
  map<uint32, UnitStatistics> byUnit = 1;
  map<uint32, UpgradeStatistics> byUpgrade = 2;
  optional uint32 clicks = 3;
}
message UnitStatistics {
  uint32 clicks = 1;
  double elapsedFirst = 2;
  Decimal num = 3;
  Decimal twinnum = 4;
}
message UpgradeStatistics {
  uint32 clicks = 1;
  double elapsedFirst = 2;
  Decimal num = 3;
}

message Options {
  optional bool isCustomTheme = 1;
  optional string theme = 2;
  optional bool showAdvancedUnitData = 3;
  // enum, but this is simpler serialization
  optional string notation = 4;
  // enum, but this is simpler serialization
  optional string velocityUnit = 5;
  // enum, but this is simpler serialization
  optional string durationFormat = 6;

  optional bool fpsAuto = 7;
  optional int32 fps = 8;
}

message AchievementsUI {
  bool earned = 1;
  bool unearned = 2;
  bool masked = 3;
  string order = 4;
  bool reverse = 5;
}

message Decimal {
  oneof value {
    float tiny = 1;
    double small = 2;
    string big = 3;
  }
}