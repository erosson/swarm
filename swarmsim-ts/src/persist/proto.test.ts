import json from "./persist.test.json";
import { pipe } from "fp-ts/lib/function";
import * as P2 from "../generated/persist";
import * as Proto from "./proto";
import fs from "fs/promises";
import LZ from "lz-string";

test("persist proto: basic", () => {
  function smallD(small: number): P2.Decimal {
    return { value: { oneofKind: "small", small } };
  }
  function bigD(big: string): P2.Decimal {
    return { value: { oneofKind: "big", big } };
  }
  function tinyD(tiny: number): P2.Decimal {
    return { value: { oneofKind: "tiny", tiny } };
  }
  const body: P2.Persist = {
    units: { 1: smallD(123) },
    upgrades: { 2: bigD("456") },
    achievements: { 3: 789 },
    watched: { 2: 1 },
    skippedMillis: 0,
    statistics: {
      byUnit: {
        1: {
          clicks: 3,
          elapsedFirst: 123456,
          num: tinyD(123),
          twinnum: tinyD(456),
        },
      },
      byUpgrade: {
        1: { clicks: 3, elapsedFirst: 123456, num: tinyD(123) },
      },
    },
    version: { started: "1.1.14", saved: "1.1.14" },
  };
  expect(
    pipe(
      body,
      (v) => P2.Persist.toJson(v),
      (v) => P2.Persist.fromJson(v)
    )
  ).toEqual(body);
});

// TODO: decimal precision comparison issues
test.skip.each(json.exports)("persist proto: $name", (d) => {
  const body = JSON.parse(d.body);

  // write files for ad-hoc size comparisons.
  // protos don't help this as much as I thought...!
  // fs.writeFile(
  // `${__dirname}/${d.name}.bin`,
  // LZ.compressToBase64(P2.Persist.toBinary(Proto.fromLegacy(body)).toString())
  // );
  // fs.writeFile(
  // `${__dirname}/${d.name}.raw`,
  // P2.Persist.toBinary(Proto.fromLegacy(body)).toString()
  // );
  // fs.writeFile(`${__dirname}/${d.name}.old`, d.export_);

  // console.log(P2.Persist.fields);
  expect(
    pipe(
      body,
      Proto.fromLegacy,
      (v) => P2.Persist.toJson(v),
      (v) => P2.Persist.fromJson(v),
      Proto.toLegacy
    )
  ).toEqual(body);
  expect(
    pipe(
      body,
      Proto.fromLegacy,
      (v) => P2.Persist.toJsonString(v),
      (v) => P2.Persist.fromJsonString(v),
      Proto.toLegacy
    )
  ).toEqual(body);
  expect(
    pipe(
      body,
      Proto.fromLegacy,
      (v) => P2.Persist.toBinary(v),
      (v) => P2.Persist.fromBinary(v),
      Proto.toLegacy
    )
  ).toEqual(body);
});
// TODO: decimal precision comparison issues
test.each(json.exports)("proto snapshot: $name", (d) => {
  const body = JSON.parse(d.body);
  expect(
    pipe(
      body,
      Proto.fromLegacy,
      (v) => P2.Persist.toBinary(v),
      (b) => [b.byteLength, btoa(b.toString())]
    )
  ).toMatchSnapshot(`binary protobuf: ${d.name}`);
});
