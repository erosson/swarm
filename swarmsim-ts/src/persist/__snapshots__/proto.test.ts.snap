// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proto snapshot: init: binary protobuf: init 1`] = `
[
  750,
  "MTAsOSw4LDEsMTgsNSwxMywwLDAsMTI4LDYzLDEwLDksOCwyLDE4LDUsMTMsMCwwLDEyLDY2LDEwLDksOCwzLDE4LDUsMTMsMCwwLDMyLDY1LDEwLDksOCw0LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNSwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDYsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw3LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsOCwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDksMTgsNSwxMywwLDAsMCwwLDEwLDksOCwxMCwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDExLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMTIsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwxMywxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDE0LDE4LDUsMTMsMCwwLDEyOCw2NCwxMCw5LDgsMTUsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwxNiwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDE3LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMTgsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwxOSwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDIwLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMjEsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwyMiwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDIzLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMjQsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwyNSwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDI2LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMjcsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwyOCwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDI5LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMzAsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwzMSwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDMyLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMzMsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwzNCwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDM1LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMzYsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwzNywxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDM4LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMzksMTgsNSwxMywwLDAsMCwwLDEwLDksOCw0MCwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDQxLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNDIsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw0MywxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDQ0LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNDUsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw0NiwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDQ3LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNDgsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw0OSwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDUwLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNTEsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw1MiwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDUzLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNTQsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw1NSwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDU2LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNTcsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw1OCwxOCw1LDEzLDAsMCwwLDAsMjYsMTEsOCwxMTksMTcsMCwwLDAsMCw4MCwxOSwzNyw2NSw0MiwyLDI0LDAsNTAsMCwxNjIsMSwxNiwxMCw2LDQ5LDQ2LDQ5LDQ2LDQ5LDUyLDE4LDYsNDksNDYsNDksNDYsNDksNTIsMTcwLDEsMTIsOCwxNTMsMTkyLDIyNywxNTYsNiwxNiwxMjgsMTU2LDE3NCwxNzMsMiwxNzgsMSwxMiw4LDE1MywxOTIsMjI3LDE1Niw2LDE2LDEyOCwxNTYsMTc0LDE3MywyLDE4NiwxLDEyLDgsMTUzLDE5MiwyMjcsMTU2LDYsMTYsMTI4LDE1NiwxNzQsMTczLDIsMTk0LDEsMTIsOCwxNTMsMTkyLDIyNywxNTYsNiwxNiwxMjgsMTU2LDE3NCwxNzMsMiwyMDIsMSwxMSw4LDIwNCwxOTcsMjI3LDE1Niw2LDE2LDE5MiwyNTEsMjAzLDEyMQ==",
]
`;

exports[`proto snapshot: nests: binary protobuf: nests 1`] = `
[
  2203,
  "MTAsOSw4LDEsMTgsNSwxMywwLDAsMTI4LDYzLDEwLDksOCwyLDE4LDUsMTMsMjQ5LDE4MywxMjIsOTgsMTAsOSw4LDMsMTgsNSwxMywxOTUsNjAsODgsNzIsMTAsOSw4LDQsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw1LDE4LDUsMTMsMjE2LDc0LDI1NCw4OSwxMCw5LDgsNiwxOCw1LDEzLDQzLDEwNCw1MCw3MCwxMCw5LDgsNywxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDgsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw5LDE4LDUsMTMsMCwwLDY0LDY0LDEwLDksOCwxMCwxOCw1LDEzLDAsMCwyNTAsNjgsMTAsOSw4LDExLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMTIsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwxMywxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDE0LDE4LDUsMTMsMCwwLDEyOCw2NCwxMCw5LDgsMTUsMTgsNSwxMywxMzAsMTU5LDE5Niw5MiwxMCw5LDgsMTYsMTgsNSwxMyw5MywxMTUsNzEsODQsMTAsOSw4LDE3LDE4LDUsMTMsNTUsMTc1LDQ1LDc3LDEwLDksOCwxOCwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDE5LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMjAsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwyMSwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDIyLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMjMsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwyNCwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDI1LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMjYsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwyNywxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDI4LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMjksMTgsNSwxMywwLDAsMCwwLDEwLDksOCwzMCwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDMxLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMzIsMTgsNSwxMywwLDAsMCwwLDEwLDksOCwzMywxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDM0LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsMzUsMTgsNSwxMywwLDAsMTYwLDY2LDEwLDksOCwzNiwxOCw1LDEzLDAsMCwxNjgsNjYsMTAsOSw4LDM3LDE4LDUsMTMsMCw0OCwyMjAsNzAsMTAsOSw4LDM4LDE4LDUsMTMsMCw1NCw1Miw3MiwxMCw5LDgsMzksMTgsNSwxMywwLDM2LDE3MSw3MSwxMCw5LDgsNDAsMTgsNSwxMywwLDExMCwxNDUsNzIsMTAsOSw4LDQxLDE4LDUsMTMsMCwxNzYsOCw3MSwxMCw5LDgsNDIsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw0MywxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDQ0LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNDUsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw0NiwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDQ3LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNDgsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw0OSwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDUwLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNTEsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw1MiwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDUzLDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNTQsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw1NSwxOCw1LDEzLDAsMCwwLDAsMTAsOSw4LDU2LDE4LDUsMTMsMCwwLDAsMCwxMCw5LDgsNTcsMTgsNSwxMywwLDAsMCwwLDEwLDksOCw1OCwxOCw1LDEzLDAsMCwwLDAsMTgsOSw4LDEsMTgsNSwxMywwLDAsMTY4LDY1LDE4LDksOCwyLDE4LDUsMTMsMCwwLDMyLDY2LDE4LDksOCwzLDE4LDUsMTMsMCwwLDEyOCw2NCwxOCw5LDgsNSwxOCw1LDEzLDAsMCwxOTIsNjQsMTgsOSw4LDYsMTgsNSwxMywwLDAsMTI4LDY0LDE4LDksOCw3LDE4LDUsMTMsMCwwLDY0LDY0LDE4LDksOCwyNSwxOCw1LDEzLDAsMCwxMjgsNjMsMTgsOSw4LDI2LDE4LDUsMTMsMCwwLDY0LDY0LDE4LDksOCwyNywxOCw1LDEzLDAsMCwxNjAsNjQsMTgsOSw4LDQ1LDE4LDUsMTMsMCwwLDEyOCw2MywxOCw5LDgsNDYsMTgsNSwxMywwLDAsMCw2NCwxOCw5LDgsNDcsMTgsNSwxMywwLDAsNjQsNjQsMTgsOSw4LDQ4LDE4LDUsMTMsMCwwLDY0LDY0LDE4LDksOCw0OSwxOCw1LDEzLDAsMCw2NCw2NCwxOCw5LDgsNTAsMTgsNSwxMywwLDAsMTI4LDY0LDE4LDksOCw1MSwxOCw1LDEzLDAsMCwxMjgsNjQsMTgsOSw4LDUyLDE4LDUsMTMsMCwwLDY0LDY0LDE4LDksOCw2NywxOCw1LDEzLDAsMCwxMjgsNjMsMTgsOSw4LDY4LDE4LDUsMTMsMCwwLDEyOCw2MywxOCw5LDgsNjksMTgsNSwxMywwLDAsMTI4LDYzLDI2LDExLDgsMSwxNywwLDAsMCwyMjIsMjUsMzgsMTc0LDY1LDI2LDExLDgsMiwxNywwLDAsMCwzMCw2NywzMywxNzQsNjUsMjYsMTEsOCwzLDE3LDAsMCwwLDEwOCw3MCw1NywxNzQsNjUsMjYsMTEsOCwxMSwxNywwLDAsMCwyNDIsMTEyLDI4LDE3NCw2NSwyNiwxMSw4LDEyLDE3LDAsMCwwLDIwMCwxNTYsMzAsMTc0LDY1LDI2LDExLDgsMTQsMTcsMCwwLDAsMjAyLDksMzAsMTc0LDY1LDI2LDExLDgsMTUsMTcsMCwwLDAsOTgsMTYyLDM4LDE3NCw2NSwyNiwxMSw4LDE3LDE3LDAsMCwwLDI0LDQ2LDQxLDE3NCw2NSwyNiwxMSw4LDE4LDE3LDAsMCwwLDI0MCw3Myw2OCwxNzQsNjUsMjYsMTEsOCwxOSwxNywwLDAsMCwxNCwxNzQsMjIzLDE3NSw2NSwyNiwxMSw4LDIwLDE3LDAsMCwwLDIxNCw4NSw4NCwxNzQsNjUsMjYsMTEsOCw4MCwxNywwLDAsMCw5OCwxMzYsMzEsMTc0LDY1LDI2LDExLDgsODYsMTcsMCwwLDAsNCwyOCw0NiwxNzQsNjUsMjYsMTEsOCw4OSwxNywwLDAsMCwxNzgsMTAyLDg1LDE3NCw2NSwyNiwxMSw4LDkyLDE3LDAsMCwwLDIyOCwxMzUsMTMyLDE3NCw2NSwyNiwxMSw4LDk1LDE3LDAsMCwwLDg0LDEyMSwxNDAsMTc0LDY1LDI2LDExLDgsOTgsMTcsMCwwLDAsMTQ2LDIyNCwyMjYsMTc1LDY1LDI2LDExLDgsMTEzLDE3LDAsMCwwLDE2NCwxNjIsNjYsMTc0LDY1LDI2LDExLDgsMTE1LDE3LDAsMCwwLDAsMCwxMjAsMTUxLDY0LDI2LDExLDgsMTE2LDE3LDAsMCwwLDEyMSwxMzksMTY3LDE3Niw2NSwyNiwxMSw4LDExNywxNywwLDAsMCwyMCw1NSw1MywxNzQsNjUsMjYsMTEsOCwxMTksMTcsMCwwLDAsMjM4LDExMSw1MywxNzQsNjUsMzQsNCw4LDUsMTYsMiwzNCw0LDgsNiwxNiwyLDM0LDQsOCw3LDE2LDIsMzQsMTMsOCwyNSwxNiwyNTUsMjU1LDI1NSwyNTUsMjU1LDI1NSwyNTUsMjU1LDI1NSwxLDM0LDEzLDgsMjYsMTYsMjU1LDI1NSwyNTUsMjU1LDI1NSwyNTUsMjU1LDI1NSwyNTUsMSwzNCw0LDgsNDUsMTYsMCwzNCw0LDgsNDYsMTYsMCwzNCw0LDgsNDcsMTYsMCwzNCw0LDgsNDgsMTYsMCwzNCw0LDgsNDksMTYsMCwzNCw0LDgsNTAsMTYsMCwzNCw0LDgsNTEsMTYsMCwzNCw0LDgsNTIsMTYsMCw0MiwxODQsNiwxMCwyOSw4LDE1LDE4LDI1LDgsMTksMTcsMCwwLDAsMjQyLDExMiwyOCwxNzQsNjUsMjYsNSwxMywwLDY0LDYyLDY4LDM0LDUsMTMsMCw2NCwxNzcsNjgsMTAsMjksOCwxNiwxOCwyNSw4LDQwLDE3LDAsMCwwLDIwMiw5LDMwLDE3NCw2NSwyNiw1LDEzLDEyOCwxMTUsMjIyLDcxLDM0LDUsMTMsMjI0LDE2NSwyMiw3MywxMCwyOSw4LDE3LDE4LDI1LDgsNDEsMTcsMCwwLDAsMjQsNDYsNDEsMTc0LDY1LDI2LDUsMTMsMTY1LDIwLDkyLDc1LDM0LDUsMTMsMjMxLDE0NSwxNTMsNzcsMTAsMjksOCwxOCwxOCwyNSw4LDUsMTcsMCwwLDAsMjE0LDg1LDg0LDE3NCw2NSwyNiw1LDEzLDAsMTU2LDQ1LDcwLDM0LDUsMTMsMCwxNTYsNDUsNzAsMTAsMjksOCwzNSwxOCwyNSw4LDIsMTcsMCwwLDAsOTgsMTM2LDMxLDE3NCw2NSwyNiw1LDEzLDAsMCwzMiw2NiwzNCw1LDEzLDAsMCwxNjAsNjYsMTAsMjksOCwzNiwxOCwyNSw4LDIsMTcsMCwwLDAsMTIwLDE2MSwzNywxNzQsNjUsMjYsNSwxMywwLDAsMTY4LDY1LDM0LDUsMTMsMCwwLDE2OCw2NiwxMCwyOSw4LDM3LDE4LDI1LDgsNywxNywwLDAsMCwxMzAsNTYsNDQsMTc0LDY1LDI2LDUsMTMsMCw0OCw5Miw2OSwzNCw1LDEzLDAsNDgsMjIwLDcwLDEwLDI5LDgsMzgsMTgsMjUsOCwzLDE3LDAsMCwwLDE3OCwxMDIsODUsMTc0LDY1LDI2LDUsMTMsMCw1NCwxODAsNzAsMzQsNSwxMywwLDU0LDUyLDcyLDEwLDI5LDgsMzksMTgsMjUsOCwxLDE3LDAsMCwwLDIyOCwxMzUsMTMyLDE3NCw2NSwyNiw1LDEzLDAsMzYsNDMsNzAsMzQsNSwxMywwLDM2LDE3MSw3MSwxMCwyOSw4LDQwLDE4LDI1LDgsNCwxNywwLDAsMCw4NCwxMjEsMTQwLDE3NCw2NSwyNiw1LDEzLDAsMzYsMTQ2LDcwLDM0LDUsMTMsMCwxMTAsMTQ1LDcyLDEwLDI5LDgsNDEsMTgsMjUsOCwxLDE3LDAsMCwwLDE0NiwyMjQsMjI2LDE3NSw2NSwyNiw1LDEzLDAsMTc2LDgsNjksMzQsNSwxMywwLDE3Niw4LDcxLDE4LDIyLDgsMSwxOCwxOCw4LDE2LDE3LDAsMCwwLDEwNiw3MCwzMSwxNzQsNjUsMjYsNSwxMywwLDAsMTY4LDY1LDE4LDIyLDgsMiwxOCwxOCw4LDMyLDE3LDAsMCwwLDMwLDY3LDMzLDE3NCw2NSwyNiw1LDEzLDAsMCwzMiw2NiwxOCwyMiw4LDMsMTgsMTgsOCw0LDE3LDAsMCwwLDEzNiw4NCw0MCwxNzQsNjUsMjYsNSwxMywwLDAsMTI4LDY0LDE4LDIyLDgsNSwxOCwxOCw4LDYsMTcsMCwwLDAsMTM4LDE1MCwzMCwxNzQsNjUsMjYsNSwxMywwLDAsMTkyLDY0LDE4LDIyLDgsNiwxOCwxOCw4LDQsMTcsMCwwLDAsMTAwLDIyMywzNSwxNzQsNjUsMjYsNSwxMywwLDAsMTI4LDY0LDE4LDIyLDgsNywxOCwxOCw4LDMsMTcsMCwwLDAsMTA0LDQ3LDQ4LDE3NCw2NSwyNiw1LDEzLDAsMCw2NCw2NCwxOCwyMiw4LDI1LDE4LDE4LDgsMSwxNywwLDAsMCwxNDYsMTA2LDMwLDE3NCw2NSwyNiw1LDEzLDAsMCwxMjgsNjMsMTgsMjIsOCwyNiwxOCwxOCw4LDMsMTcsMCwwLDAsNjQsMTA5LDQxLDE3NCw2NSwyNiw1LDEzLDAsMCw2NCw2NCwxOCwyMiw4LDI3LDE4LDE4LDgsNSwxNywwLDAsMCwyMjQsMTM3LDg0LDE3NCw2NSwyNiw1LDEzLDAsMCwxNjAsNjQsMTgsMjIsOCw0NSwxOCwxOCw4LDEsMTcsMCwwLDAsMTk4LDEyNywzMSwxNzQsNjUsMjYsNSwxMywwLDAsMTI4LDYzLDE4LDIyLDgsNDYsMTgsMTgsOCwxLDE3LDAsMCwwLDM4LDEyMSwzNywxNzQsNjUsMjYsNSwxMywwLDAsMCw2NCwxOCwyMiw4LDQ3LDE4LDE4LDgsMSwxNywwLDAsMCwxNjAsNDMsNDQsMTc0LDY1LDI2LDUsMTMsMCwwLDY0LDY0LDE4LDIyLDgsNDgsMTgsMTgsOCwxLDE3LDAsMCwwLDE3NiwzOCw0NywxNzQsNjUsMjYsNSwxMywwLDAsNjQsNjQsMTgsMjIsOCw0OSwxOCwxOCw4LDEsMTcsMCwwLDAsMTAsMTM4LDg1LDE3NCw2NSwyNiw1LDEzLDAsMCw2NCw2NCwxOCwyMiw4LDUwLDE4LDE4LDgsMiwxNywwLDAsMCwxNDIsMTE3LDEzMiwxNzQsNjUsMjYsNSwxMywwLDAsMTI4LDY0LDE4LDIyLDgsNTEsMTgsMTgsOCwyLDE3LDAsMCwwLDQ2LDE3NCwxNDAsMTc0LDY1LDI2LDUsMTMsMCwwLDEyOCw2NCwxOCwyMiw4LDUyLDE4LDE4LDgsMSwxNywwLDAsMCwxNzgsMzIsMjI4LDE3NSw2NSwyNiw1LDEzLDAsMCw2NCw2NCwxOCwyMiw4LDY3LDE4LDE4LDgsMSwxNywwLDAsMCwxNjQsMTYyLDY2LDE3NCw2NSwyNiw1LDEzLDAsMCwxMjgsNjMsMTgsMjIsOCw2OCwxOCwxOCw4LDEsMTcsMCwwLDAsNjAsNCwxMzEsMTc0LDY1LDI2LDUsMTMsMCwwLDEyOCw2MywxOCwyMiw4LDY5LDE4LDE4LDgsMSwxNywwLDAsMCwyMCwxNjgsMjIyLDE3NSw2NSwyNiw1LDEzLDAsMCwxMjgsNjMsMjQsMjEyLDEsNTAsMTIsOCwwLDE4LDYsMTAwLDk3LDExNCwxMDcsMTA4LDEyMSwyNCwxLDU4LDE1LDgsMSwxNiwxLDI0LDEsMzQsNywxMDAsMTAxLDEwMiw5NywxMTcsMTA4LDExNiwxNjIsMSwxNiwxMCw2LDQ5LDQ2LDQ5LDQ2LDQ5LDUyLDE4LDYsNDksNDYsNDksNDYsNDksNTIsMTcwLDEsMTEsOCwxOTIsMTQ5LDIxMSwxNTYsNiwxNiwxMjgsMTk1LDEzMywxMDIsMTc4LDEsMTEsOCwxOTIsMTQ5LDIxMSwxNTYsNiwxNiwxMjgsMTk1LDEzMywxMDIsMTg2LDEsMTIsOCwxOTQsMTkxLDIyNywxNTYsNiwxNiwxOTIsMTcxLDE3NSwxNjcsMSwxOTQsMSwxMSw4LDE5MiwxNDksMjExLDE1Niw2LDE2LDEyOCwxOTUsMTMzLDEwMiwyMDIsMSwxMiw4LDE4MywxNTYsMjI4LDE1Niw2LDE2LDE5MiwxNzgsMjMyLDE2OSwzLDIxMCwxLDExLDgsMTM3LDE5MCwyMjcsMTU2LDYsMTYsMTkyLDE1OCwxMzQsNzE=",
]
`;
